### PR TITLE
ExperimentsSynced, SyncErrors, and SyncErrorRate CloudWatch metrics

### LINKED ISSUE #355: ExperimentsSynced, SyncErrors, and SyncErrorRate CloudWatch metrics
### Bug Description                                                                                                                
The sync job's ExperimentsSynced, SyncErrors, and SyncErrorRate CloudWatch metrics are never published despite the sync job processing experiments successfully. The PersistentSyncFailure metric (same job) and cleanup job metrics (DataCleanupCount, CleanupErrors) publish correctly, confirming CloudWatch connectivity and IAM permissions are fine.

The root cause is likely that _publish_metrics() at sync_job.py:655 silently fails (line 702-703 catches all exceptions and only logs a warning), or that the metric was published but is being queried without the right parameters. However, since the cleanup job uses an identical pattern and works, the most likely cause is that _publish_metrics is silently swallowing an error
   - possibly related to the "Timestamp": get_current_datetime() call or the SyncErrorRate Percent unit type, which differs from the cleanup job's simpler Count-only metrics.

### Steps to Reproduce
  1. Publish an experiment (status becomes pending)
  2. Wait for sync job to run (every 5 minutes) - confirm via DB that local_sync_status changes to synced
  3. Query CloudWatch: aws cloudwatch list-metrics --namespace "OptiNiSt/BackgroundJobs"
  4. Observe ExperimentsSynced, SyncErrors, SyncErrorRate are absent from the metric list

### Expected Behavior
 After the sync job processes experiments, ExperimentsSynced, SyncErrors, and SyncErrorRate should appear in CloudWatch under OptiNiSt/BackgroundJobs, matching the cleanup job's behavior.

### Actual Behavior
Only PersistentSyncFailure, CleanupErrors, and DataCleanupCount appear. The three sync metrics are never published. The exception handler at sync_job.py:702-703 may be silently swallowing errors with only a logger.warning.

### Environment
  - AWS (deployed infrastructure)
  - sync_job.py: PublishedExperimentSyncJob._publish_metrics() (line 655)
  - Cleanup job metrics work correctly with identical pattern

### Priority
  Low - The DB local_sync_status field is the authoritative source for sync state and works correctly. CloudWatch metrics are supplementary observability. No data loss or functional impact.

### Screenshots/Logs
`$ aws cloudwatch list-metrics --namespace "OptiNiSt/BackgroundJobs"`
- Metrics returned: PersistentSyncFailure, CleanupErrors, DataCleanupCount
- Missing: ExperimentsSynced, SyncErrors, SyncErrorRate

### Additional Comments
  - To debug: change logger.warning at line 703 to logger.error with exc_info=True to capture the actual exception
  - The PersistentSyncFailure metric publishes from the same class, same boto3 pattern, confirming IAM/network is fine
  - One difference: _publish_metrics includes a Percent unit type (SyncErrorRate) that the cleanup job doesn't use - worth  investigating if that causes the entire put_metric_data call to fail
  - No unit tests exist for _publish_metrics in the sync job

### PR DESCRIPTION
**Pull request by milesAraya**
#### Summary
- The sync job's ExperimentsSynced, SyncErrors, and SyncErrorRate CloudWatch metrics were never published despite experiments being processed successfully.
- Root cause: the dataview endpoint marks experiments as `synced` when users view them, so by the time the sync job runs, `_get_pending_experiments()` returns empty and the method returns early — skipping `_publish_metrics()` entirely.
- Secondary issue: all three metrics were published in a single `put_metric_data` call, so a failure in the `Percent`-unit SyncErrorRate metric would silently block the two `Count` metrics. The exception handler used `logger.warning` without `exc_info`, hiding the actual error.

#### Design Decisions
- **Publish metrics on the empty-pending path** -- Added `_publish_metrics(0, 0)` before the early return at line 174. This ensures metrics appear in CloudWatch every run (every 5 minutes), even when the dataview endpoint has already handled all syncing. Publishing zeros is cheap and provides continuous proof-of-life for the sync job. The cleanup job's metrics work because `users_to_cleanup` is typically non-empty, so it never hits its equivalent early return.
- **Split `put_metric_data` into two calls (Count then Percent)** -- CloudWatch rejects the entire batch if any single `MetricDatum` is invalid. Isolating SyncErrorRate into its own call ensures ExperimentsSynced and SyncErrors are committed before the Percent metric is attempted. Trade-off: one extra API call per run (~$0.00 cost impact).
- **Single timestamp variable** -- The original code called `get_current_datetime()` three times (once per metric), producing slightly different timestamps. Now captures one timestamp and reuses it for consistency.
- **Explicit `float` for error_rate** -- Changed `0` to `0.0` and added `float()` cast to ensure the Percent value is always a float, avoiding any potential type-related CloudWatch rejection.
- **Upgraded `logger.warning` to `logger.error` with `exc_info=True`** -- The original handler silently swallowed exceptions with no traceback, making diagnosis impossible. Now captures the full stack trace.

### Evidence
- `pytest studio/tests/app/common/core/background/test_sync_job.py` -- 36 tests pass (13 new)
- `test_publishes_zero_metrics_when_no_pending` directly validates the root cause fix

### References
- CloudWatch `put_metric_data` batch semantics: entire call fails if any MetricDatum is invalid
- Dataview self-healing path: `studio/app/common/routers/dataview.py:293-298` sets `local_sync_status = synced` when data is locally available, bypassing the sync job

#### Files changed
- `studio/app/common/core/background/sync_job.py` -- Publish metrics on empty-pending early return; split put_metric_data into Count and Percent calls; use consistent timestamp; ensure error_rate is float; upgrade error logging from warning to error with traceback
- `studio/tests/app/common/core/background/test_sync_job.py` -- Add TestValidationLogicMetrics (4 tests: empty pending, correct counts, gather exceptions, all-fail) and TestPublishMetrics (9 tests: metric structure, zero counts, float type, 100% error rate, error logging, no-raise guarantee, Percent isolation, consistent timestamp, namespace)

### Manual Testcases
- Deploy to staging and publish an experiment (status becomes `pending`)
- View the experiment in the UI before the sync job runs (triggers dataview self-healing path)
- Wait for sync job to run (every 5 minutes)
- Query CloudWatch: `aws cloudwatch list-metrics --namespace "OptiNiSt/BackgroundJobs"`
- Verify ExperimentsSynced, SyncErrors, and SyncErrorRate now appear (with value 0 if dataview handled the sync)
- Confirm PersistentSyncFailure, DataCleanupCount, and CleanupErrors still publish correctly
- Run `pytest studio/tests/app/common/core/background/test_sync_job.py -v` -- all 36 tests pass

### Unit, Integration, Contract Test Coverage
- 13 new unit tests covering both `_run_validation_logic` metric publishing paths and `_publish_metrics` edge cases
- Tests verify behavior against the original bugs: early return without metrics, silent exception swallowing, batch failure propagation

#### Difficulties (if any)
The root cause was not in `_publish_metrics` itself but in the caller never reaching it. The dataview endpoint's self-healing behavior (`dataview.py:293-298`) creates a race where experiments are marked `synced` before the sync job runs, leaving `_get_pending_experiments()` empty on every invocation.

#### Risk Assessment
| Area | Risk | Notes |
|------|------|-------|
| Metric publishing on empty pending | Low | Publishes zeros — no functional impact, adds observability |
| Split put_metric_data calls | Low | One extra CloudWatch API call per run; no behavioral change to metric data |
| Error logging upgrade | Low | Changes log level from warning to error for metric failures; no functional impact |

### CODE REVIEW

File: studio/app/common/core/background/sync_job.py [hunk 1]
```diff
@@ -663,7 +664,8 @@ def _publish_metrics(cls, synced_count: int, error_count: int):
             cloudwatch: "CloudWatchClient" = boto3.client("cloudwatch")
 
             total = synced_count + error_count
-            error_rate = (error_count / total * 100) if total > 0 else 0
+            error_rate = float(error_count / total * 100) if total > 0 else 0.0
+            timestamp = get_current_datetime()
 
             cloudwatch.put_metric_data(
```
Comment (itutu-tienday) [inline studio/app/common/core/background/sync_job.py:670]: This question may not directly relate to the fix in this PR, but it's relevant, so I'd like to ask.
(I could do an AI review, but I'll ask directly first.)

In `sync_job.py`, I believe `def _do_proactive_download_sync` requests data synchronization to the public dataview page via the API (ALB).

Here, the public page may be running in multiple containers. (A free user would access the public page on a free instance, and a premium user would access the public page on a premium instance.)

However, it seems that `def _do_proactive_download_sync` only issues one API request per record.

Is this behavior correct?
*is the synchronization request propagated to all public page containers?
*Or does it mean that requests are limited to a single container, but there are essentially no issues with state management? (Perhaps just a slight decrease in performance?)

Comment (milesAraya) [reply studio/app/common/core/background/sync_job.py:670]: I see your point. While there may be a slight performance decrease, this would be minor as thumbnails ae very lightweight and are downloaded on demand if not present:
  1. Startup sync is the main download step. `run_startup_sync downloads` all missing published experiments when a container starts. 
  2. The proactive download is an optimization, not a state synchronization mechanism. The DB state `local_sync_status` is shared via RDS, all containers see the same status. The proactive download just pre-caches thumbnails/metadata on the local filesystem so users don't wait for an S3 download on first access.                                                                                                                   
  3. Containers that miss the proactive sync will download on-demand. The `download_single_experiment` endpoint itself skips work if files already exist locally. Containers without the cached files will fetch from S3 when a user actually requests the data, very slightly slower first load.

However, seeing as the `feature/public-tier-instance` branch introduces a dedicated public-site tier with its own target group, ASG, and ECS service. With the public tier having typically 1 container (unlikely to reach 2, although technically could), the "one request hits one container" problem mostly disappears.

Comment (itutu-tienday) [reply studio/app/common/core/background/sync_job.py:670]: I understand the current specifications.

Furthermore, if this function is to be optimized in the future, it is likely that the following methods will be used to accurately propagate synchronization requests to each container:
1. The background container issues a synchronization request to all public web running containers.
2. Alternatively, the public web running container detects the synchronization request flag and performs the synchronization itself.

At this point, I understand that there appear to be no problems with the operation of the current specifications.

**Review by itutu-tienday — APPROVED** [review]
